### PR TITLE
ENH: Add `spin` command `gdb`; customize `docs` and `test`

### PR DESCRIPTION
**Pull request by stefanv**
Partially addresses #24080

### CODE REVIEW

**Comment by stefanv** [comment]
AttributeError: module 'importlib' has no attribute 'util'

Hrm, what's going on with that Python?

**Comment by stefanv** [comment]
Should be fixed in https://github.com/scientific-python/spin/pull/76

**Comment by charris** [comment]
Close/reopen

**Comment by stefanv** [comment]
I'll have to make a new release of spin before this will work.

**Comment by charris** [comment]
LGTM, should we wait for the new spin release?

**Comment by stefanv** [comment]
Yes, I'm working on adding the `run` command (dealing with some Windows intricacies), and then I think we should be set.

**Comment by stefanv** [comment]
I've expanded the PR a bit, and it now does the following:

- Use the new `spin docs` command, but customize it for NumPy with an option to install dependencies.
- Customize the `spin test` command to add `-m` (marker) and `-j` flags (parallel runners).
  Now that we have an example of how to do it, it would be easy to add others.
- Add the new `spin run` command. This sets PYTHONPATH and invokes the given command.
- Add a NumPy command: `spin gdb` for running code through gdb.

**Comment by stefanv** [comment]
Failures seem unrelated.

**Comment by mattip** [comment]
It seems the macos_arm64 failure is related. When running [`spin test`](https://github.com/numpy/numpy/blob/2f375c0f9f19085684c9712d602d22a2b4cb4c8e/tools/ci/cirrus_macosx_arm64.yml#L57), the [error](https://github.com/numpy/numpy/pull/24083/checks?check_run_id=14710242472) is
```
$ /Users/admin/numpy-dev/bin/python3.10 -m pytest \
   --rootdir=<long-path>/build-install/usr/lib/python3.10/site-packages \
   -n auto -m 'not slow' numpy
ERROR: usage: __main__.py [options] [file_or_dir] [file_or_dir] [...]
__main__.py: error: unrecognized arguments: -n numpy
  inifile: /private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cirrus-ci-build/pytest.ini
  rootdir: /private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cirrus-ci-build/build-install/usr/lib/python3.10/site-packages
```

**Comment by stefanv** [comment]
Missing pytest-xdist

File: .spin/cmds.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    metavar='MARKEXPR',
+    default="not slow",
+    help="Run tests with the given markers"
+)
+@click.option(
+    "-j",
+    "n_jobs",
+    metavar='N_JOBS',
+    default="auto",
+    help="Number of parallel jobs for testing; use all CPUs by default"
```
Comment (mattip) [inline .spin/cmds.py:76]: I think running on all cores should be an opt-in, not a default option.

**Review by mattip — COMMENTED** [review]
Lokos good. One nit about using all cores by default. We should change the docs build job (on cirecle CI) to use `spin doc`, but that can be a separate PR.

File: .spin/cmds.py (hunk 1 above)
Comment (stefanv) [reply .spin/cmds.py]: Set the default to "1", and only pass the "-j" flag when changed. This also means that `pytest-xdist` is not required by default, which is probably a good thing.

**Comment by stefanv** [comment]
Also added `-t` (`--tests`) for selecting tests, and `-v` (`--verbose`).

**Comment by stefanv** [comment]
I'll probably add the flag additions to spin 0.5, then we can further reduce the custom command here.

File: test_requirements.txt [hunk 2]
```diff
@@ -6,6 +6,7 @@ hypothesis==6.24.1
 pytest==6.2.5
 pytz==2021.3
 pytest-cov==3.0.0
+pytest-xdist
```
Comment (mattip) [inline test_requirements.txt:9]: Do you want to keep this as a project-wide requirement?

**Review by mattip — COMMENTED** [review]
We can merge this as-is if you want, since it pins to `spin==0.4` and then update again for a newer `spin`, or would you rather wait?

File: test_requirements.txt (hunk 2 above)
Comment (stefanv) [reply test_requirements.txt:9]: Probably, since otherwise the -j option doesn't work. But I don't feel particularly strongly about it either way.

**Comment by stefanv** [comment]
I think 0.5 won't be for a while, so let's get this in first.

**Comment by charris** [comment]
Thanks Stefan.

**Pull request merged**